File: sdk-rust/examples/integration.rs
///! Integration example: list markets, check balance, place a trade, check balance again.
///!
///! Usage:
///!   export DECIBEL_PRIVATE_KEY="0x..."
///!   export DECIBEL_ACCOUNT_ADDRESS="0x..."
///!   export APTOS_NODE_API_KEY="..."          # optional, for higher rate limits
///!   cargo run --example integration
use decibel_sdk::client::read::DecibelReadClient;
use decibel_sdk::client::write::{DecibelWriteClient, PlaceOrderArgs};
use decibel_sdk::config::testnet_config;
use decibel_sdk::models::{TimeInForce, VolumeWindow};
use std::env;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let config = testnet_config();
    let api_key = env::var("APTOS_NODE_API_KEY").ok();
    let private_key = env::var("DECIBEL_PRIVATE_KEY")
        .expect("Set DECIBEL_PRIVATE_KEY env var");
    let account_address = env::var("DECIBEL_ACCOUNT_ADDRESS")
        .expect("Set DECIBEL_ACCOUNT_ADDRESS env var");

    // ── 1. Read client ─────────────────────────────────────────────────
    let read = DecibelReadClient::new(config.clone(), api_key.clone(), None)?;

    // ── List all available markets ──────────────────────────────────────
    println!("=== Available Markets ===");
    let markets = read.get_all_markets().await?;
    for m in &markets {
        println!(
            "  {} (addr: {}…)  max_leverage: {}x  tick: {}  lot: {}",
            m.market_name,
            &m.market_addr[..10],
            m.max_leverage,
            m.tick_size,
            m.lot_size,
        );
    }
    println!("Total markets: {}\n", markets.len());

    // Pick the first market for the rest of the demo.
    let market = markets
        .first()
        .expect("No markets found — is the testnet deployment live?");
    let market_name = &market.market_name;

    // ── Fetch current price ─────────────────────────────────────────────
    let prices = read.get_market_price_by_name(market_name).await?;
    if let Some(p) = prices.first() {
        println!(
            "=== {market_name} Prices ===\n  mark: {}  mid: {}  oracle: {}  funding: {} bps\n",
            p.mark_px, p.mid_px, p.oracle_px, p.funding_rate_bps,
        );
    }

    // ── Fetch order-book depth ──────────────────────────────────────────
    let depth = read.get_market_depth(market_name, Some(5)).await?;
    println!("=== {market_name} Order Book (top 5) ===");
    println!("  Bids:");
    for b in &depth.bids {
        println!("    {} @ {}", b.size, b.price);
    }
    println!("  Asks:");
    for a in &depth.asks {
        println!("    {} @ {}", a.size, a.price);
    }
    println!();

    // ── 2. Write client ─────────────────────────────────────────────────
    let write = DecibelWriteClient::new(
        config,
        &private_key,
        &account_address,
        false, // skip_simulate
        false, // no_fee_payer
        api_key,
        None,  // gas_price_manager
        None,  // time_delta_ms
    )?;

    let subaccount = write.get_primary_subaccount_addr();
    println!("=== Account ===");
    println!("  address:    {account_address}");
    println!("  subaccount: {subaccount}\n");

    // ── Check balance BEFORE the trade ──────────────────────────────────
    let before = read
        .get_account_overview(&subaccount, Some(VolumeWindow::ThirtyDays), None)
        .await?;
    println!("=== Balance BEFORE Trade ===");
    println!("  equity:       {}", before.perp_equity_balance);
    println!("  margin:       {}", before.total_margin);
    println!("  unrealised:   {}", before.unrealized_pnl);
    println!("  withdrawable: {}\n", before.usdc_cross_withdrawable_balance);

    // ── Place a small limit order well below market (won't fill) ────────
    let limit_price = if let Some(p) = prices.first() {
        (p.mid_px * 0.90 * 100.0).round() / 100.0   // 10 % below mid
    } else {
        1.0
    };
    let order_size = market.min_size;

    println!("=== Placing Order ===");
    println!(
        "  market: {market_name}  side: BUY  price: {limit_price}  size: {order_size}  tif: GTC"
    );

    let result = write
        .place_order(PlaceOrderArgs {
            market_name: market_name.clone(),
            price: limit_price,
            size: order_size,
            is_buy: true,
            time_in_force: TimeInForce::GoodTillCanceled,
            is_reduce_only: false,
            client_order_id: Some("rust-example-001".into()),
            stop_price: None,
            tp_trigger_price: None,
            tp_limit_price: None,
            sl_trigger_price: None,
            sl_limit_price: None,
            builder_addr: None,
            builder_fee: None,
            subaccount_addr: Some(subaccount.clone()),
            tick_size: Some(market.tick_size),
        })
        .await?;

    if result.success {
        println!("  ✓ order_id: {:?}", result.order_id);
        println!("  tx_hash:    {:?}\n", result.transaction_hash);
    } else {
        println!("  ✗ error: {:?}\n", result.error);
    }

    // ── Check balance AFTER the trade ───────────────────────────────────
    let after = read
        .get_account_overview(&subaccount, Some(VolumeWindow::ThirtyDays), None)
        .await?;
    println!("=== Balance AFTER Trade ===");
    println!("  equity:       {}", after.perp_equity_balance);
    println!("  margin:       {}", after.total_margin);
    println!("  unrealised:   {}", after.unrealized_pnl);
    println!("  withdrawable: {}\n", after.usdc_cross_withdrawable_balance);

    // ── Show open orders ────────────────────────────────────────────────
    let open = read.get_user_open_orders(&subaccount).await?;
    println!("=== Open Orders ({}) ===", open.len());
    for o in &open {
        println!(
            "  {} {} {} @ {} (remaining: {})",
            o.order_id,
            if o.is_buy { "BUY" } else { "SELL" },
            o.market,
            o.price,
            o.remaining_size,
        );
    }

    println!("\nDone.");
    Ok(())
}

File: sdk-python/examples/integration.py
"""End-to-end integration example for the Decibel Python SDK.

This script demonstrates:
  1. Reading market data (REST)
  2. Streaming live prices (WebSocket)
  3. Placing an order (Write client)

Usage:
    # Read-only (no API key required for public data):
    uv run python examples/integration.py

    # With trading (requires API key and private key):
    DECIBEL_API_KEY=... DECIBEL_PRIVATE_KEY=... uv run python examples/integration.py
"""

from __future__ import annotations

import asyncio
import os

from decibel_sdk import (
    CandlestickInterval,
    DecibelReadClient,
    DecibelWriteClient,
    PlaceOrderArgs,
    TimeInForce,
    mainnet_config,
)


async def read_market_data() -> None:
    """Fetch public market data via REST."""
    config = mainnet_config()

    async with DecibelReadClient(config) as client:
        # List all available markets
        markets = await client.get_all_markets()
        print(f"Available markets ({len(markets)}):")
        for m in markets[:5]:
            print(f"  {m.market_name}  leverage={m.max_leverage}x  tick={m.tick_size}")

        if not markets:
            return

        # Get current prices
        prices = await client.get_all_market_prices()
        print(f"\nCurrent prices ({len(prices)}):")
        for p in prices[:5]:
            print(f"  {p.market}  mark={p.mark_px}  oracle={p.oracle_px}")

        # Get order book depth for the first market
        first = markets[0].market_name
        depth = await client.get_market_depth(first, limit=5)
        print(f"\n{first} order book (top 5):")
        for bid in depth.bids[:5]:
            print(f"  BID  {bid.price}  x{bid.size}")
        for ask in depth.asks[:5]:
            print(f"  ASK  {ask.price}  x{ask.size}")

        # Get recent candlesticks
        import time

        now = int(time.time())
        candles = await client.get_candlesticks(
            first,
            CandlestickInterval.ONE_HOUR,
            start_time=now - 86400,
            end_time=now,
        )
        print(f"\n{first} 1h candles (last 24h): {len(candles)} bars")
        if candles:
            last = candles[-1]
            print(f"  Latest: O={last.open} H={last.high} L={last.low} C={last.close}")


async def stream_prices() -> None:
    """Stream live market prices via WebSocket."""
    config = mainnet_config()

    async with DecibelReadClient(config) as client:
        print("\nStreaming BTC-USD prices (5 ticks)...")
        async with client.subscribe_market_price("BTC-USD") as prices:
            count = 0
            async for price in prices:
                print(f"  mark={price.mark_px}  mid={price.mid_px}  oi={price.open_interest}")
                count += 1
                if count >= 5:
                    break


async def place_order_example() -> None:
    """Place a limit order (requires API key and private key)."""
    api_key = os.environ.get("DECIBEL_API_KEY")
    private_key = os.environ.get("DECIBEL_PRIVATE_KEY")

    if not api_key or not private_key:
        print("\nSkipping order placement (set DECIBEL_API_KEY and DECIBEL_PRIVATE_KEY)")
        return

    config = mainnet_config()
    account_address = os.environ.get("DECIBEL_ACCOUNT_ADDRESS", "")
    if not account_address:
        print("\nSkipping order placement (set DECIBEL_ACCOUNT_ADDRESS)")
        return

    writer = DecibelWriteClient(
        config=config,
        private_key_hex=private_key,
        account_address=account_address,
        api_key=api_key,
    )
    reader = DecibelReadClient(config, api_key=api_key)

    try:
        # Get market info for tick/lot sizes
        market = await reader.get_market_by_name("BTC-USD")

        args = PlaceOrderArgs(
            market_name="BTC-USD",
            is_buy=True,
            size=market.min_size,
            price=30000.0,
            time_in_force=TimeInForce.GOOD_TILL_CANCELED,
            is_reduce_only=False,
        )

        result = await writer.place_order(args)
        if result.success:
            print(f"\nOrder placed! ID: {result.order_id}, TX: {result.transaction_hash}")
        else:
            print(f"\nOrder failed: {result.error}")
    finally:
        await reader.close()


async def main() -> None:
    print("=== Decibel Python SDK Integration Example ===\n")

    await read_market_data()
    await stream_prices()
    await place_order_example()

    print("\nDone!")


if __name__ == "__main__":
    asyncio.run(main())
